// gatsby-cli/lib/reporter/reporters/ink/components/error.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _path = _interopRequireDefault(require("path"));

var _ink = require("ink");

var _lodash = require("lodash");

const File = ({
  filePath,
  location
}) => {
  const lineNumber = (0, _lodash.get)(location, `start.line`);
  let locString = ``;

  if (lineNumber) {
    locString += `:${lineNumber}`;
    const columnNumber = (0, _lodash.get)(location, `start.column`);

    if (columnNumber) {
      locString += `:${columnNumber}`;
    }
  }

  return _react.default.createElement(_ink.Color, {
    blue: true
  }, _path.default.relative(process.cwd(), filePath), locString);
};

const DocsLink = ({
  docsUrl
}) => {
  // TODO: when there's no specific docsUrl, add helpful message describing how
  // to submit an issue
  if (docsUrl === `https://gatsby.dev/issue-how-to`) return null;
  return _react.default.createElement(_ink.Box, {
    marginTop: 1
  }, "See our docs page for more info on this error: ", docsUrl);
};

const Error = ({
  details
}) => // const stackLength = get(details, `stack.length`, 0
_react.default.createElement(_ink.Box, {
  marginY: 1,
  flexDirection: "column"
}, _react.default.createElement(_ink.Box, {
  flexDirection: "column"
}, _react.default.createElement(_ink.Box, {
  flexDirection: "column"
}, _react.default.createElement(_ink.Box, null, _react.default.createElement(_ink.Box, {
  marginRight: 1
}, _react.default.createElement(_ink.Color, {
  black: true,
  bgRed: true
}, ` ${details.level} `, details.id ? `#${details.id} ` : ``), _react.default.createElement(_ink.Color, {
  red: true
}, details.type ? ` ` + details.type : ``))), _react.default.createElement(_ink.Box, {
  marginTop: 1
}, details.text), details.filePath && _react.default.createElement(_ink.Box, {
  marginTop: 1
}, "File:", ` `, _react.default.createElement(File, {
  filePath: details.filePath,
  location: details.location
}))), _react.default.createElement(DocsLink, {
  docsUrl: details.docsUrl
})));

var _default = Error;
exports.default = _default;
